cowbird/services/impl/geoserver.py

```diff
@@ -22,11 +27,49 @@
 LOGGER = get_logger(__name__)
 
 
-class Geoserver(Service):
+def geoserver_response_handling(func):
+    """
+    Decorator for response and logging handling for the different Geoserver HTTP requests.
+
+    @param func : Function executing a http request to Geoserver
+    @return : Response object
+    """
+
+    @functools.wraps(func)
+    def wrapper(*args, **kwargs):
+        # Geoserver responses are often full HTML pages for codes 400-499, so text/content is omitted from
+        # the logs. Error responses in the 500-599 range are usually concise, so their text/content were included
+        # in the logs to help eventual debugging.
+        response = func(*args, **kwargs)
+        operation = func.__name__
+        response_code = response.status_code
+        regex_to_find = "Workspace &#39;.*&#39; already exists"
+        if response_code in (200, 201):
+            LOGGER.info("Operation [%s] was successful.", operation)
+        elif response_code == 401 and re.search(regex_to_find, response.text):
+            # This is done because Geoserver's reply/error code is misleading in this case and
+            # returns HTML content.
+            LOGGER.error("Geoserver workspace already exists")
+        elif response_code == 401:
+            LOGGER.error("Operation [%s] failed because it lacks valid authentication credentials.",
+                         operation)
+        elif response_code == 403 and operation == "_remove_workspace_request":
+            LOGGER.error(
+                "Geoserver workspace [%s] is not empty. Make sure `recurse` is set to `true` to delete workspace")
+        elif response_code == 500:
+            LOGGER.error("Operation [%s] failed : %s", operation, response.text)
+        else:
+            LOGGER.error("Operation [%s] failed with HTTP error code [%s]", operation, response_code)
```

Comment:
I guess this is where a 404 not found (aka server is not running) would fall... We should throw something so that the autoretry can kick-in and send the request again in a couple of seconds. 
It seems that whatever happens with geoserver requests, we just swallow the error, log something and continue as if nothing happened. 
If the error is temporary, throwing it will cause the request to be retried. 
If the error is fatal, throw a recognizable exception so that geoserver's celery task can call `RequestTask.abort_chain` and avoid any downstream operations.

Reply:
Good find.

The other cases (401, 403 and most of the 500) are fatal errors not worth retrying as they are mostly `workspace/datastore already/doesn't exist` and retrying with the same parameters will only yield the same error.

Reply:
Ok, I was testing this and turns out 404 is also used by Geoserver when a ressource is not found (like trying to delete a workspace that does not exist), in which case retrying won't solve anything. I'll have to expand this decorator a bit to catch those cases and prevent useless long retries

Reply:
In fact, I was wrong with the 404, I thing that a timed-out request will yield a 504 Gateway Time-out

Reply:
Could also be a ReadTimeout 408 depending on how the client/server handle it.

Reply:
After some testing, what I see when the Geoserver is offline is the actual request failing (say `requests.get(...)` )  and raising an exception, so no error codes or anything gets returned. The task then auto-retries (because `RequestException` is in `RequestTask`'s auto-retry list) but it doesn't show up in the logs.

For the rest, I don't think there is a lot of "retry" potential, they're fatal errors that would hint at an errors in the code, duplicate users or  bad inputs.

I'll be wrapping the fatal errors in an Exception and raise it to stop the chain, and I'll work something up so we can get some feedback when Geoserver is unavailable.